### PR TITLE
Imrprove error handling & docs

### PR DESCRIPTION
**Pull request by pmarks**
- Propagate errors a bit better: if there's an IO error, the send() method of ShardSender will return a generic error and ShardWriter::finish() will return the original error
- Improve documentation

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -418,42 +485,49 @@ where
...
+        let sent = self.buf_tx.send((buf, done));
+        if sent.is_err() {
+            return Err(format_err!("writer thread shut down, see ShardWriterThread.finish() for error"));
+        }
 
         match r {
-            Ok(r) => return r,
-            _ => (),
+            Ok(r) => return Ok(r),
+            Err(_) => (),
```
Comment (nlhepler) [inline src/lib.rs]: This should match on TryRecvError::Empty with (), and raise the error for any other type.

**Comment by nlhepler** [comment]
Outside of my one comment, the only thing missing is a test.

This is reminding me I need to finish polishing up my branch with subkey typing. I'll rebase that on this when you land it.

File: src/lib.rs (hunk 1 above)
Comment (pmarks) [reply src/lib.rs]: thanks, fixed.

**Comment by pmarks** [comment]
So I have tested this by injecting errors manually in the file reader code (see lib.rs:171) .  Since its using libc directly and not a trait, it's hard to make this stategy into a unit test.  I added a test that truncates the output file & makes sure we gracefully get an error.

**Pull request merged**